### ISSUE TITLE
cmd/go: 'go mod download' fails with xyz+incompatible replacement targets while 'go mod verifies' works

### ISSUE THREAD

**Issue by dmitris** [golang/go#34254]
### What version of Go are you using (`go version`)?
go version go1.13 darwin/amd64
</pre>

### Does this issue reproduce with the latest release?
yes

### What operating system and processor architecture are you using (`go env`)?
GOHOSTARCH="amd64"
GOHOSTOS="darwin"

### What did you do?
my `go.mod` file contains replacement targets like:
```
github.com/pkg/errors => git.ourcompany.com/mirror-github/pkg--errors v0.8.1+incompatible
gopkg.in/yaml.v2 => git.ourcompany.com/mirror-github/go-yaml--yaml v2.2.2+incompatible
```

I ran `go mod verify` and `go mod download`.

### What did you expect to see?
Since the project used to build with go1.12 and still builds / tests run fine, I expect both commands to work without errors.

### What did you see instead?
`go mod verify` shows `all modules verified` (as expected), but `go mod download` fails with:
```
$ go mod download
go: finding git.ourcompany.com/mirror-github/pkg--errors v0.8.1+incompatible
go: finding git.ourcompany.com/mirror-github/go-yaml--yaml v2.2.2+incompatible
git.ourcompany.com/mirror-github/pkg--errors@v0.8.1+incompatible: invalid version: +incompatible suffix not allowed: major version v0 is compatible
git.ourcompany.com/mirror-github/go-yaml--yaml@v2.2.2+incompatible: invalid version: +incompatible suffix not allowed: module contains a go.mod file, so semantic import versioning is required
```

This seems related to #34217; on slack @bcmills said that it is likely a bug.

>  would argue that the acceptable versions should be:
`gopkg.in/yaml.v2 => git.ourcompany.com/mirror-github/go-yaml--yaml/v2 v2.2.2`
(because `[…]/v2` should be able to replace `gopkg.in/[…].v2`, bug notwithstanding) or
`gopkg.in/yaml.v2 => git.ourcompany.com/mirror-github/go-yaml--yaml v2.2.2`
(because the go.mod file should say gopkg.in/yaml.v2 explicitly, and that should be good enough for validation).

**Comment by bcmills** [comment]
The first entry here:
```
	github.com/pkg/errors => git.ourcompany.com/mirror-github/pkg--errors v0.8.1+incompatible
```

really is incorrect, and the `go` command is correct to flag it as such. The version `v0.8.1` is compatible with the module paths on *both* sides, so it very clearly should not have a `+incompatible` suffix.

**Comment by bcmills** [comment]
The second entry here:
```
	gopkg.in/yaml.v2 => git.ourcompany.com/mirror-github/go-yaml--yaml v2.2.2+incompatible
```
is also incorrect, because the module has a `go.mod` file: the module path should have a `/v2` suffix, or a `gopkg.in` path and a `.v2` suffix.

**Comment by bcmills** [comment]
What _will_ work, as of [CL 206761](https://golang.org/cl/206761), is:

```
	gopkg.in/yaml.v2 => git.ourcompany.com/mirror-github/go-yaml--yaml/v2 v2.2.2
```

**Comment by bcmills** [comment]
This should be fixed at head.

The fix likely patches cleanly onto Go 1.13, but I do not believe that it meets our backporting criteria: I have only been able to find two affected modules (`gopkg.in/yaml.v2` and `gopkg.in/src-d/go-git.v4`), and there is at least one possible workaround.

One possible workaround is to check out the replacement source into a local directory, and target the `replace` directive to the local directory rather than the `git`-hosted repo. There may be others.

**Comment by imjuanleonard** [comment]
Work for me after removing the +incompatible suffix, thanks @bcmills

Go version 1.13.4

GOPRIVATE="company.repo.com"
GOPROXY="https://company.golang.artifact.com,direct"
GOSUMDB="https://sum.golang.org"

**Comment by dmitris** [comment]
works for me as well with the tip Go with this replace:
```
	gopkg.in/yaml.v2 => git.ourcompany.com/mirror-github/go-yaml--yaml/v2 v2.2.2
```

> The fix likely patches cleanly onto Go 1.13, but I do not believe that it meets our backporting criteria

that's sad (for those of us affected 😄 ) - maybe an exception could be considered since `gopkg.in/yaml.v2` is the most popular yaml parsing library?  It is the first hit in the package search results for "yaml" - https://pkg.go.dev/search?q=yaml, shows 8654 known imports and likely there are many ones in the corporate codebases as well. @bcmills

**Comment by bcmills** [comment]
> maybe an exception could be considered since `gopkg.in/yaml.v2` is the most popular yaml parsing library?

We've literally had only the two reports of this issue. (I suspect that most people are not carrying their own patches to `gopkg.in/yaml.v2`, and thus can use a proxy rather than a `replace` directive.)
